File: code/jvm/im-domain/src/main/kotlin/im/domain/messages/MessageValidator.kt
package im.domain.messages

import im.domain.Either
import im.domain.failure
import im.domain.success

class MessageValidator {
    companion object {
        private const val MIN_LENGTH = 1
        private const val MAX_LENGTH = 300
    }

    fun validate(message: String): Either<List<MessageValidationError>, Unit> {
        val errors = mutableListOf<MessageValidationError>()

        if (message.isBlank()) {
            errors.add(MessageValidationError.ContentBlank)
        }

        if (message.length !in MIN_LENGTH..MAX_LENGTH) {
            errors.add(MessageValidationError.ContentLength(MIN_LENGTH, MAX_LENGTH))
        }

        if (errors.isNotEmpty()) {
            return failure(errors)
        }

        return success(Unit)
    }
}

File: code/js/src/Domain/messages/MessageValidator.ts
import { MessageValidationError } from './MessageValidationError';

export class MessageValidator {
    private static readonly MIN_LENGTH = 1;
    private static readonly MAX_LENGTH = 300;

    constructor(
        private readonly minLength: number = MessageValidator.MIN_LENGTH,
        private readonly maxLength: number = MessageValidator.MAX_LENGTH,
    ) {}

    validate(message: string): MessageValidationError[] {
        const errors: MessageValidationError[] = [];

        if (message.trim().length === 0) {
            errors.push(MessageValidationError.ContentBlank);
        }

        if (message.length < this.minLength || message.length > this.maxLength) {
            errors.push(new MessageValidationError.ContentLength(this.minLength, this.maxLength));
        }

        if (errors.length > 0) {
            return errors;
        }

        return [];
    }
}
